src/Native/LibTorchSharp/THSNN.cpp

```diff
@@ -499,19 +499,34 @@ Tensor THSNN_poisson_loss(const Tensor input, const Tensor target, const bool lo
     )
 }
 
-Optimizer THSNN_Adam_ctor(const Tensor* parameters, const int length, const double learnig_rate)
+Optimizer THSNN_Adagrad_ctor(const Tensor* parameters, const int length, const double learning_rate, const double lr_decay, const double weight_decay)
+{
+    auto params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
+    auto options = torch::optim::AdagradOptions(learning_rate).lr_decay(lr_decay).weight_decay(weight_decay);
+
+    return new std::shared_ptr<torch::optim::Optimizer>(std::make_shared<torch::optim::Adagrad>(torch::optim::Adagrad(params, options)));
+}
+
+Optimizer THSNN_Adam_ctor(const Tensor* parameters, const int length, const double learning_rate)
+{
+    auto  params = toTensors<at::Tensor>((torch::Tensor**)parameters, length);
```

Comment:
There is an AdamOptions we should be using here https://pytorch.org/cppdocs/api/structtorch_1_1optim_1_1_adam_options.html?highlight=adamoptions#_CPPv4N5torch5optim11AdamOptionsE

Reply:
Also this has numerous additional parameters available on it

Reply:
Yes, we can add that. There are a lot of options that are being skipped for the various optimizers. I can add them systematically.

Reply:
Thanks.  If you skip anything knowingly please make sure there's a note of some form in an open bug e.g. just add it to one of the existing bugs listing missing things